### PR TITLE
Abbreviate cheatsheet

### PR DESCRIPTION
**Pull request by ArenaL5**
Refactor of code in #2994 as @actionless requested. I also sent a second commit to add documentation to the newly exposed API (and tweak a couple sentences), please tell me back if I need to edit or remove it.

(By the way, the argument `keygroup` should be added to the documentation for both constructors `awful.key`, but I decided against doing it myself partly because the documentation for the whole page is a tad confusing and I'm afraid to break something.)

### CODE REVIEW

File: lib/awful/hotkeys_popup/widget.lua [hunk 1]
```diff
@@ -55,7 +55,7 @@ local widget = {
 widget.hide_without_description = true
 
 --- Merge hotkey records into one if they have the same modifiers and
--- description.
+-- description. Records with five or more keys will abbreviate them.
```
Comment (actionless) [inline lib/awful/hotkeys_popup/widget.lua]: please mention what it's only for awful.keys but not manually added hotkeys for the apps

File: lib/awful/hotkeys_popup/widget.lua [hunk 2]
```diff
@@ -285,6 +295,53 @@ function widget.new(args)
...
+        -- applicable: they do not need this method.
+        for _, keys in pairs(self._cached_awful_keys) do
+            for _, params in pairs(keys) do
+                if #params.keylist > 4 then
+                    -- assuming here keygroups will never overlap;
+                    -- if they ever do, another for loop will be necessary:
+                    local keygroup = gtable.find_first_key(self._keygroups, function(_, v)
+                        return not not gtable.hasitem(v, params.keylist[1])
+                    end)
+                    local first, last, count, tally = nil, nil, 0, {}
```
Comment (actionless) [inline lib/awful/hotkeys_popup/widget.lua]: i'm not sure, mb you could use `#tally` instead of `count`?

Comment (ArenaL5) [reply lib/awful/hotkeys_popup/widget.lua]: It's not reliable. [The manual for 5.3 is verbose but pretty clear:](https://www.lua.org/manual/5.3/manual.html#3.4.7), `#table` may return any key `n` as long as:
* `n` ≥ 0 and integer,
* `table[n]` exists,
* and `table[n+1]` is nil.

If `tally[1]` doesn't exist (if, for example, `2/3/4/5/6` is checked against `numrow`) or there are gaps in the table, `#tally` has undefined behaviour. ([Lua 5.1](https://www.lua.org/manual/5.1/manual.html#2.5.5) explicitly says that if `tally[1]` is nil, `#tally` **may** be 0)

File: lib/awful/hotkeys_popup/widget.lua (hunk 1 above)
Comment (ArenaL5) [reply lib/awful/hotkeys_popup/widget.lua]: Oh, right, sorry.

Just pushed another commit. Is 2d3f7e1e94b60d34cadf3e5ec531eca6884f38c6 okay?

Comment (actionless) [reply lib/awful/hotkeys_popup/widget.lua]: yup, thanks!

File: lib/awful/hotkeys_popup/widget.lua (hunk 2 above)
Comment (actionless) [reply lib/awful/hotkeys_popup/widget.lua]: yeah, it's indeed not good:

```lua
> t = {}
> #t
0
> t[2] = 123
> #t
0
> t[1] = 123
> #t
2
```

**Comment by ArenaL5** [comment]
Thank you both.

**Pull request merged**